### ISSUE TITLE
[ENOMEM] make hb-subset compatible with low memory environment allocator

### ISSUE THREAD

**Issue by ebraminio** [harfbuzz/harfbuzz#2568]
Apply 

```diff
diff --git a/test/fuzzing/hb-subset-fuzzer.cc b/test/fuzzing/hb-subset-fuzzer.cc
index cc535c86..115a6356 100644
--- a/test/fuzzing/hb-subset-fuzzer.cc
+++ b/test/fuzzing/hb-subset-fuzzer.cc
@@ -67,7 +67,7 @@ trySubset (hb_face_t *face,
 
 extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
 {
-  /* alloc_state = size; TODO: enable it */ /* see src/failing-alloc.c */
+  alloc_state = size; /* see src/failing-alloc.c */
 
   hb_blob_t *blob = hb_blob_create ((const char *)data, size,
                                    HB_MEMORY_MODE_READONLY, nullptr, nullptr);
```

over #2566 and run `CXXFLAGS="-fsanitize=address,fuzzer-no-link" meson build --default-library=static -Dfuzzer_ldflags="-fsanitize=address,fuzzer" && ninja -Cbuild test/fuzzing/hb-{shape,draw,subset,set}-fuzzer && build/test/fuzzing/hb-subset-fuzzer`

```
==4122==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 48 byte(s) in 1 object(s) allocated from:
    #0 0x55f8d61cd101 in calloc (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x2a5101)
    #1 0x55f8d62a9872 in hb_calloc_impl ~/harfbuzz/build/../src/failing-alloc.c:46:31
    #2 0x55f8d620615e in hb_blob_t* hb_object_create<hb_blob_t>() ~/harfbuzz/build/../src/hb-object.hh:227:24
    #3 0x55f8d6205d51 in hb_blob_create ~/harfbuzz/build/../src/hb-blob.cc:82:16
    #4 0x55f8d6206d32 in hb_blob_create_sub_blob ~/harfbuzz/build/../src/hb-blob.cc:144:10
    #5 0x55f8d620d269 in _hb_face_for_data_reference_table(hb_face_t*, unsigned int, void*) ~/harfbuzz/build/../src/hb-face.cc:178:21
    #6 0x55f8d620e318 in hb_face_t::reference_table(unsigned int) const ~/harfbuzz/build/../src/hb-face.hh:77:12
    #7 0x55f8d620e1b5 in hb_face_reference_table ~/harfbuzz/build/../src/hb-face.cc:378:16
    #8 0x55f8d66c5b13 in _subset_table(hb_subset_plan_t*, unsigned int) ~/harfbuzz/build/../src/hb-subset.cc:224:31
    #9 0x55f8d66c4e4b in hb_subset ~/harfbuzz/build/../src/hb-subset.cc:256:17
    #10 0x55f8d6205866 in trySubset(hb_face_t*, unsigned int const*, int, bool, bool, bool) ~/harfbuzz/build/../test/fuzzing/hb-subset-fuzzer.cc:36:23
    #11 0x55f8d62054d4 in trySubset(hb_face_t*, unsigned int const*, int, unsigned char const*) ~/harfbuzz/build/../test/fuzzing/hb-subset-fuzzer.cc:64:3
    #12 0x55f8d620514c in LLVMFuzzerTestOneInput ~/harfbuzz/build/../test/fuzzing/hb-subset-fuzzer.cc:88:3
    #13 0x55f8d610511e in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x1dd11e)
    #14 0x55f8d6107350 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x1df350)
    #15 0x55f8d6107ce9 in fuzzer::Fuzzer::MutateAndTestOne() (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x1dfce9)
    #16 0x55f8d610a097 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x1e2097)
    #17 0x55f8d60f3ba8 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x1cbba8)
    #18 0x55f8d60e12d3 in main (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x1b92d3)
    #19 0x7f36c54ad001 in __libc_start_main (/usr/lib/libc.so.6+0x27001)

Indirect leak of 48 byte(s) in 1 object(s) allocated from:
    #0 0x55f8d61cd101 in calloc (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x2a5101)
    #1 0x55f8d62a9872 in hb_calloc_impl ~/harfbuzz/build/../src/failing-alloc.c:46:31
    #2 0x55f8d620615e in hb_blob_t* hb_object_create<hb_blob_t>() ~/harfbuzz/build/../src/hb-object.hh:227:24
    #3 0x55f8d6205d51 in hb_blob_create ~/harfbuzz/build/../src/hb-blob.cc:82:16
    #4 0x55f8d620509f in LLVMFuzzerTestOneInput ~/harfbuzz/build/../test/fuzzing/hb-subset-fuzzer.cc:72:21
    #5 0x55f8d610511e in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x1dd11e)
    #6 0x55f8d6107350 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x1df350)
    #7 0x55f8d6107ce9 in fuzzer::Fuzzer::MutateAndTestOne() (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x1dfce9)
    #8 0x55f8d610a097 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x1e2097)
    #9 0x55f8d60f3ba8 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x1cbba8)
    #10 0x55f8d60e12d3 in main (~/harfbuzz/build/test/fuzzing/hb-subset-fuzzer+0x1b92d3)
    #11 0x7f36c54ad001 in __libc_start_main (/usr/lib/libc.so.6+0x27001)

SUMMARY: AddressSanitizer: 96 byte(s) leaked in 2 allocation(s).
INFO: to ignore leaks on libFuzzer side use -detect_leaks=0.

MS: 1 EraseBytes-; base unit: 9d991a978fa5aa07af23be02b7de2ba77142b236
0x0,0x1,0x0,0x0,0x0,0x1,0x0,0xd6,0x1e,0x0,0x0,0x0,0x40,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x1,0x0,0x2d,0x0,0x82,0x1,0xed,0x82,0x0,0x40,0x82,0x82,0xed,
\x00\x01\x00\x00\x00\x01\x00\xd6\x1e\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00-\x00\x82\x01\xed\x82\x00@\x82\x82\xed
artifact_prefix='./'; Test unit written to ./leak-139679df2901d396ead97dc9506bcadcb76beb17
Base64: AAEAAAABANYeAAAAQAAAAAAAAAAAAAAAAAABAC0AggHtggBAgoLt
```

Have applied

```diff
diff --git a/src/hb-subset-plan.cc b/src/hb-subset-plan.cc
index b16f5077..f4720d4c 100644
--- a/src/hb-subset-plan.cc
+++ b/src/hb-subset-plan.cc
@@ -146,7 +146,7 @@ static inline void
     gpos->collect_variation_indices (&c);
 
   gdef->remap_layout_variation_indices (layout_variation_indices, layout_variation_idx_map);
-  
+
   gdef.destroy ();
   gpos.destroy ();
 }
@@ -314,7 +314,9 @@ hb_subset_plan_t *
 hb_subset_plan_create (hb_face_t         *face,
                       hb_subset_input_t *input)
 {
-  hb_subset_plan_t *plan = hb_object_create<hb_subset_plan_t> ();
+  hb_subset_plan_t *plan;
+  if (unlikely (!(plan = hb_object_create<hb_subset_plan_t> ())))
+    return const_cast<hb_subset_plan_t *> (&Null (hb_subset_plan_t));
 
   plan->drop_hints = input->drop_hints;
   plan->desubroutinize = input->desubroutinize;
diff --git a/src/hb-subset.cc b/src/hb-subset.cc
index 31127392..b1676d29 100644
--- a/src/hb-subset.cc
+++ b/src/hb-subset.cc
@@ -261,6 +261,7 @@ hb_subset (hb_face_t *source, hb_subset_input_t *input)
 end:
 
   hb_face_t *result = success ? hb_face_reference (plan->dest) : hb_face_get_empty ();
+  result = result ? result : hb_face_get_empty ();
   hb_subset_plan_destroy (plan);
   return result;
 }
diff --git a/test/fuzzing/hb-subset-fuzzer.cc b/test/fuzzing/hb-subset-fuzzer.cc
index cc535c86..115a6356 100644
--- a/test/fuzzing/hb-subset-fuzzer.cc
+++ b/test/fuzzing/hb-subset-fuzzer.cc
@@ -67,7 +67,7 @@ trySubset (hb_face_t *face,
 
 extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
 {
-  /* alloc_state = size; TODO: enable it */ /* see src/failing-alloc.c */
+  alloc_state = size; /* see src/failing-alloc.c */
 
   hb_blob_t *blob = hb_blob_create ((const char *)data, size,
                                    HB_MEMORY_MODE_READONLY, nullptr, nullptr);
```

for starters but subset API aren't written having that in mind.

**Comment by ebraminio** [comment]
@garretrieger this reveals an interesting aspect of subset APIs and guess is worth to look.

**Comment by garretrieger** [comment]
Got the fuzzer running with the low memory allocator enabled. I'll begin chasing down all the problems it finds.

**Comment by garretrieger** [comment]
For hb_subset_plan_t I think we'll want to add a "successful" member (like hb_set_t for example). So that we can detect when it's from the null pool. Then have hb_subset(...) be a noop if given a plan in an error state.

**Comment by garretrieger** [comment]
https://github.com/harfbuzz/harfbuzz/pull/2594 fixes the above memory leaks. Also ran the fuzzer for a few minutes and it didn't find anything else.

**Comment by ebraminio** [comment]
Fixed by #2594

Thank you very much @garretrieger :)
